include/seqan3/range/views/kmer_hash.hpp

```diff
@@ -290,16 +281,66 @@ class kmer_hash_view<urng_t>::shape_iterator
-        shape_{s_}, text_left{it_start}, text_right{std::ranges::next(it_start, s_.size(), it_end)}
+        shape_{s_}, text_left{it_start}, text_right{std::ranges::next(text_left, shape_.size() - 1, it_end)}
     {
         assert(std::ranges::size(shape_) > 0);
 
-        if (shape_.size() <= std::ranges::distance(text_left, text_right))
+        // shape size = 3
+        // Text:      1 2 3 4 5 6 7 8 9
+        // text_left: ^
+        // text_right:    ^
+        // distance(text_left, text_right) = 2
+        if (shape_.size() <= std::ranges::distance(text_left, text_right) + 1)
         {
             roll_factor = pow(sigma, static_cast<size_t>(std::ranges::size(shape_) - 1));
+            hash_full();
+        }
+    }
 
+    /*!\brief Construct from a given iterator on the text and a seqan3::shape.
+    * /param[in] it_start Iterator pointing to the first position of the text.
+    * /param[in] it_end   Sentinel pointing to the end of the text.
+    * /param[in] s_       The seqan3::shape that determines which positions participate in hashing.
+    * /param[in] is_end   Indicates that this iterator should point to the end of the text.
+    *
+    * \details
+    *
+    * If we have a common_range as underlying range, we want to preserve this property.
+    * This means that we need to have a shape_iterator that can act as end for the kmer_hash_view, i.e.
+    * the text_right iterator is equal to the end iterator of the underlying range.
+    * However, we still need to do some initialisation via hash_full:
+    * When using `std::views::reverse`, we start iterating from the end and decrement the iterator.
+    * After calling hash_full, we need to reset our text_right iterator to point to the end again.
+    *
+    * Another difference to the other constructor is that we need to do some work to determine the position of
+    * the text_left iterator. Note that we use `std::ranges::next` instead of `std::ranges::prev` because the latter
+    * only works for bidirectional ranges.    *
+    *
+    * ### Complexity
+    *
+    * Linear in size of shape.
+    */
+    shape_iterator(it_t it_start, sentinel_t it_end, shape s_, bool SEQAN3_DOXYGEN_ONLY(is_end)) : shape_{s_}
+    {
+        assert(std::ranges::size(shape_) > 0);
+
+        auto urange_size = std::ranges::distance(it_start, it_end);
+        auto step = (shape_.size() + 1 > urange_size) ? 0 : urange_size - shape_.size() + 1;
+        text_left = std::ranges::next(it_start, step, it_end);
+        text_right = it_end;
+
+        // shape size = 3
+        // Text:      1 2 3 4 5 6 7 8 9
+        // text_left: ^
+        // text_right:    ^
+        // distance(text_left, text_right) = 2
+        if (shape_.size() <= std::ranges::distance(text_left, text_right) + 1)
+        {
+            roll_factor = pow(sigma, static_cast<size_t>(std::ranges::size(shape_) - 1));
             hash_full();
         }
+
+        text_right = it_end;
```

Comment:
Does `hash_full` change the text_right, or why are you doing the assignment here again?

Reply:
Yes, it changes the iterators and text_right would point to the last element of the text, instead to the end